### PR TITLE
Opportunities in currency other than org default not auto closing

### PR DESCRIPTION
**Pull request by mdeleon0**
Round up before comparing when multicurrency is set up.
# Critical Changes
# Changes
# Issues Closed
# New Metadata
# Deleted Metadata

### CODE REVIEW

File: src/classes/PMT_PaymentCreator.cls [hunk 1]
```diff
@@ -394,11 +394,22 @@ public class PMT_PaymentCreator {
...
         Boolean amountPositive = (newOpp.Amount > 0);
         Boolean isNotAndWasNotClosed = (!oldOpp.IsClosed && !newOpp.IsClosed);
-        Boolean amountEqualPaymentReceived = (newOpp.Amount == totalPaymentAmount);
+        Boolean amountEqualPaymentReceived;
+
+        // If the org has multicurrency setted up we need to round up the currency value to avoid bugs.
+        if (UserInfo.isMultiCurrencyOrganization()) {
+            String oppCurrency = (String)newOpp.get('CurrencyIsoCode');
+            Integer decimalPlaces = UTIL_CurrencyCache.getInstance().getCurrDecimalPlaces(oppCurrency);
+            amountEqualPaymentReceived = isEqualWhenRounded(newOpp.Amount, totalPaymentAmount, decimalPlaces);
```
[the remark this answers is not in the document]
Comment (mdeleon0) [reply src/classes/PMT_PaymentCreator.cls]: Agreed.

File: src/classes/PMT_PaymentCreator.cls [hunk 2]
```diff
@@ -394,11 +394,22 @@ public class PMT_PaymentCreator {
      */
     private void closeOpportunityIfAllPaymentsReceived(Opportunity oldOpp, Opportunity newOpp, String closedStageName,
             Decimal totalPaymentAmount) {
+
         Boolean amountUpdated = (oldOpp.Amount != newOpp.Amount);
         Boolean paymentAmountReceivedUpdated = (oldOpp.npe01__Payments_Made__c != totalPaymentAmount);
         Boolean amountPositive = (newOpp.Amount > 0);
         Boolean isNotAndWasNotClosed = (!oldOpp.IsClosed && !newOpp.IsClosed);
-        Boolean amountEqualPaymentReceived = (newOpp.Amount == totalPaymentAmount);
+        Boolean amountEqualPaymentReceived;
```
[the remark this answers is not in the document]
Comment (mdeleon0) [reply src/classes/PMT_PaymentCreator.cls]: Ok.

File: src/package.xml [hunk 3]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
[the remark this answers is not in the document]
Comment (mdeleon0) [reply src/package.xml]: Not really I think this was autogenerated at some point, changing it back.

File: src/classes/PMT_PaymentCreator.cls [hunk 4]
```diff
@@ -412,6 +423,20 @@ public class PMT_PaymentCreator {
...
 
+
+    /*******************************************************************************************************
+    * @description return true if booth values are equal when rounded on the same decimal scale.
+    * @param Decimal amount1 one of the decimal values to compare.
+    * @param Decimal amount2 the other decimal value to compare.
+    * @param Integer decimalPlaces for booth values
+    * @return Boolean stating id booth Decimal values are equal.
+    */
+    private static Boolean isEqualWhenRounded(Decimal amount1, Decimal amount2, Integer decimalPlaces) {
```
[the remark this answers is not in the document]
Comment (mdeleon0) [reply src/classes/PMT_PaymentCreator.cls]: Added check for booth amount1 and amount2.

**Comment by lmeerkatz** [comment]
**lurch:attach W-028941

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x80000007fORlAAM" target="_blank">W-028941</a>

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 5]
```diff
@@ -1033,4 +1033,125 @@ private with sharing class PMT_PaymentCreator_TEST {
...
+            StageName = openStage
+        );
+
+        insert opp;
+
+        npe01__OppPayment__c payment = [SELECT id,
+                                               npe01__paid__c,
+                                               npe01__payment_amount__c,
+                                               npe01__Payment_Date__c,
+		                                       CurrencyIsoCode
```
[the remark this answers is not in the document]
Comment (lmeerkatz) [reply src/classes/PMT_PaymentCreator_TEST.cls]: There are a couple of other static references to CurrencyIsoCode you'll need to address as well.

[the remark this answers is not in the document]
Comment (mdeleon0) [reply src/classes/PMT_PaymentCreator_TEST.cls]: @lmeerkatz Dynamic queries fix the problem of retrieving the field if it is not there, however we still need to assign a value to the Currency Iso Code for the Opportunity. For what I've seen in other Unit Tests the general solution is to add

 if (!UserInfo.isMultiCurrencyOrganization()) {
    return;
}

**Comment by lmeerkatz** [comment]
When I started debugging, it looked like the test code was never being hit. After some refactoring and a lot of debug statements, I believe we've got a meaningful test (though it will only run in multicurrency orgs with more than one currency defined).

screen shot 2018-11-05 at 4 45 17 pm

File: src/classes/PMT_PaymentCreator_TEST.cls [hunk 6]
```diff
@@ -1033,4 +1033,141 @@ private with sharing class PMT_PaymentCreator_TEST {
...
+
+        UTIL_CustomSettingsFacade.getContactsSettings().Payments_Auto_Close_Stage_Name__c = closedWonStage;
+
+        //create test data
+        Account acct = new Account (
+            Name='Test Account'
+        );
+        insert acct;
+
+        // I have an opportunity for CAD100 in a stage of Prospecting.
```
Comment (lmeerkatz) [inline src/classes/PMT_PaymentCreator_TEST.cls]: Update comment - this is 100 in org default currency

**Pull request closed without merging**